### PR TITLE
Add logging for X-Request-Id header to enable deep microservice tracing

### PR DESCRIPTION
**Pull request by avatarneil**
## 📚 Purpose
There is currently no information added to log messages which allows tracing of a particular request. This PR makes it such that logs in mds-logger will have the X-Request-Id header propegated to all log messages in the context of a given HTTP request, so we can trace logs through our different microservices. This header is set currently by istio, but could also be set by something like nginx. 

## 📦 Impacts:
[mds-api-server]
[mds-logger]
[everything]

## ✅ PR Checklist
- [x] Clean up prototype code (make more composable, etc)
- [x] Test in k8s

### CODE REVIEW

File: packages/mds-api-server/middleware/http-context.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import httpContext from 'express-http-context'
+import express from 'express'
+import { ApiRequest, ApiResponse } from '../@types'
+
+export const HttpContextMiddleware = () => httpContext.middleware
+
+export const RequestIdMiddleware = () => (req: ApiRequest, res: ApiResponse, next: express.NextFunction) => {
+  httpContext.set('x-request-id', res.get('x-request-id'))
+  return next()
+}
```
Comment (avatarneil) [inline packages/mds-api-server/middleware/http-context.ts:10]: Probably ought to just merge these into one middleware, would avoid us having to double-init at the top of the file

File: packages/mds-logger/index.ts [hunk 2]
```diff
@@ -29,8 +31,9 @@ const log = (level: LogLevel, ...args: unknown[]): string[] => {
   if (redacted.length) {
     const timestamp = Date.now()
     const ISOTimestamp = new Date(timestamp).toISOString()
+    const requestId = httpContext.get('x-request-id')
```
Comment (avatarneil) [inline packages/mds-logger/index.ts:34]: We don't want to log out an `undefined` if this doesn't exist, think of a better way to do that

Comment (drtyh2o) [inline packages/mds-logger/index.ts:34]: What happens when `mds-logger` is used outside of an http request context?

Comment (avatarneil) [reply packages/mds-logger/index.ts:34]: Just logs `undefined`. I already marked that as something for me to fix, this is just the result of super hacky prototyping (hence being in draft)

File: packages/mds-logger/index.ts [hunk 3]
```diff
@@ -29,8 +31,9 @@ const log = (level: LogLevel, ...args: unknown[]): string[] => {
...
 
-    logger[level](level.toUpperCase(), ISOTimestamp, timestamp, ...redacted)
+    logger[level](level.toUpperCase(), ISOTimestamp, timestamp, requestId,...redacted)
```
Comment (drtyh2o) [inline packages/mds-logger/index.ts:36]: Does this work?

```suggestion
    logger[level](level.toUpperCase(), ISOTimestamp, timestamp, ...(requestId ? [requestId] : []), ...redacted)
```

File: packages/mds-api-server/middleware/http-context.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import httpContext from 'express-http-context'
+import express from 'express'
+import { ApiRequest, ApiResponse } from '../@types'
+
+export const HttpContextMiddleware = () => httpContext.middleware
+
+export const RequestIdMiddleware = () => (req: ApiRequest, res: ApiResponse, next: express.NextFunction) => {
+  httpContext.set('x-request-id', res.get('x-request-id'))
```
Comment (drtyh2o) [inline packages/mds-api-server/middleware/http-context.ts:8]: Should we assign an id if the gateway didn't?
```suggestion
  httpContext.set('x-request-id', res.get('x-request-id') ?? uuid())
```

**Comment by drtyh2o** [comment]
So, if a request is made to an API and the API makes an RPC service call (via HTTP POST), that service call will have a different request id value, correct? Is there a way to connect them?

**Comment by avatarneil** [comment]
@drtyh2o Yes, we can propagate the header. It will have to be done manually

File: packages/mds-api-server/middleware/http-context.ts (hunk 4 above)
Comment (avatarneil) [reply packages/mds-api-server/middleware/http-context.ts:8]: Yeah, I suppose we could, but I worry that there may be a time when the gateway _doesn't_ add it, and we don't realize that until we're tracing down a request and expecting to see that ID appear in the proxy logs... Issue would be more obvious if we don't have the ID included at all unless it's set by the gateway IMO

File: packages/mds-logger/index.ts (hunk 2 above)
Comment (avatarneil) [reply packages/mds-logger/index.ts:34]: As of recent cleanup commits, now doesn't log it at all if non-existant

File: packages/mds-api-server/middleware/http-context.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ * and includes it in contextual logs for a given request.
+ */
+const RequestIdMiddleware = (req: ApiRequest, res: ApiResponse, next: express.NextFunction) => {
+  httpContext.set('x-request-id', req.get('x-request-id'))
+  return next()
+}
+
+export const HttpContextMiddlewares = () => {
+  return [HttpContextMiddleware, RequestIdMiddleware]
+}
```
Comment (drtyh2o) [inline packages/mds-api-server/middleware/http-context.ts:18]: Don't really love the pluralized `Middlewares` ... How about:
```suggestion
export const HttpContextMiddleware = () => {
  return [httpContext.middleware, RequestIdMiddleware]
}
```

File: packages/mds-api-server/middleware/http-context.ts (hunk 4 above)
Comment (invertigo) [reply packages/mds-api-server/middleware/http-context.ts]: Given that, maybe set a static warning string?  "WARNING - No X-Request-ID found" something like that?

Comment (avatarneil) [reply packages/mds-api-server/middleware/http-context.ts]: Yeah, I can add something like that.

**Pull request merged**